packages/scripts/src/upgrade-tests/162/1_Propose.ts

```diff
@@ -1,85 +1,344 @@
+// NODE_URL_137=<OPTIMISM-NODE-URL> \
+// \
+// OPTIMISTIC_ORACLE_V2_42161=<ARBITRUM-OOV2-ADDRESS> \
+// NODE_URL_42161=<OPTIMISM-NODE-URL> \
+// \
+// OPTIMISTC_ORACLE_V2=<MAINNET-OOV2-ADDRESS> \
+// \
+// yarn hardhat run ./src/upgrade-tests/162/1_Propose.ts  --network localhost
 
 const hre = require("hardhat");
+import * as path from "path";
+import fs from "fs";
 
 const { RegistryRolesEnum } = require("@uma/common");
-
-import { Finder, Governor, Proposer, Registry } from "@uma/contracts-node/typechain/core/ethers";
-import { getContractInstance } from "../../utils/contracts";
+import { BigNumberish } from "@ethersproject/bignumber";
+import { BytesLike } from "@ethersproject/bytes";
+import {
+  ArbitrumParentMessenger,
+  Finder,
+  Governor,
+  GovernorHub,
+  GovernorRootTunnel,
+  Proposer,
+  Registry,
+} from "@uma/contracts-node/typechain/core/ethers";
+import { BaseContract, PopulatedTransaction, Signer } from "ethers";
+import { getContractInstance, getContractInstanceByUrl } from "../../utils/contracts";
 
 // PARAMETERS
 const proposerWallet = "0x2bAaA41d155ad8a4126184950B31F50A1513cE25";
 const deployed_optimistic_oracle_address = process.env["OPTIMISTC_ORACLE_V2"];
 
 const OPTIMISTIC_ORACLE_V2 = "OptimisticOracleV2"; // TODO use interfaceName.OptimisticOracle
 
-async function main() {
-  if (!deployed_optimistic_oracle_address) throw new Error("OPTIMISTC_ORACLE_V2 environment variable not set");
+// CONSTANTS
+const POLYGON_ID = 137;
+const ARBITRUM_ID = 42161;
+
+// Env vars
+const NODE_URL_ENV = "NODE_URL_";
+const OPTIMISTIC_ORACLE_V2_ENV = "OPTIMISTIC_ORACLE_V2_";
+
+export interface RelayProposal {
+  to: string;
+  value: BigNumberish;
+  data: BytesLike;
+  chainId: BigNumberish;
+}
+export interface RelayRecords {
+  block: number;
+  governorRootTunnel: RelayProposal[];
+  governorHub: RelayProposal[];
+}
+
+const relayGovernanceRootTunnelMessage = async (
+  targetAddress: string,
+  tx: PopulatedTransaction,
+  governorRootTunnel: GovernorRootTunnel
+): Promise<{
+  to: string;
+  value: BigNumberish;
+  data: BytesLike;
+}> => {
+  if (!tx.data) throw new Error("Transaction has no data");
+  const relayGovernanceData = await governorRootTunnel.populateTransaction.relayGovernance(targetAddress, tx.data);
+  console.log("- relayGovernanceData", relayGovernanceData);
+  const relay = await governorRootTunnel.populateTransaction.relayGovernance(targetAddress, tx.data);
+  const relayMessage = relay.data;
+  if (!relayMessage) throw new Error("Relay message is empty");
+  return { to: governorRootTunnel.address, value: 0, data: relayMessage };
+};
+
+const relayGovernanceHubMessage = async (
+  targetAddress: string,
+  tx: PopulatedTransaction,
+  governorHub: GovernorHub,
+  chainId: BigNumberish
+): Promise<{
+  to: string;
+  value: BigNumberish;
+  data: BytesLike;
+}> => {
+  if (!tx.data) throw new Error("Transaction has no data");
+  const calls = [{ to: targetAddress, data: tx.data }];
```

Comment:
OOC, why did we decide to do all of these in separate cross chain calls rather than grouping all of them into the same txn?

Right now, we have a separate call for addMember, registerContract, removeMember, and setImplementation address in the finder.

I doubt this is a big deal, but there are two minor issues with the non-atomic approach:
1. Bridges generally process/send txns in order in the happy path, but I don't think this is a guarantee (for instance, one could have too little gas included and might need to be retried, while the one after does not, so they would be executed out of order).
2. If they are all done in the same L2 transaction, we can be assured that they either all succeed or all fail. If we do them separately, you could have the last 2 fail, leaving the contracts in an intermediate state. We could also get weird behavior if the first fails but the last 3 are still attempted. For instance, we could have the OO get added to the finder, but not be approved by the registry.

Reply:
Good point @mrice32; I agree that these calls should be grouped by chain id to allow for atomic relays.

Because of time concerns, we finally run the script yesterday as it is now. I've left a TODO that I'll fix when we run the proposal's actions so we don't lose sight of the original script.